contributors/devel/sig-instrumentation/migration-to-structured-logging.md

```diff
@@ -0,0 +1,346 @@
+   klog.Infof("%s %s: (%v) %v%v%v [%s %s]", ...)
+}
+```
+should be changed to
+```go
+func LogHTTP(r *http.Request) {
+   klog.InfoS("Received HTTP request", ...)
+}
+```
+
+### Name arguments
+
+Even though new structured logging functions have very similar function prototype `func (string, ...interface{})` it
+has different meaning for variadic arguments. Instead of just passing arguments, now we are passing key value pairs of
+argument name and argument value. This means when migrating a log call we need to add an additional string before each
+argument, that will be used as it's name.
+
+For example
+```go
+func LogHTTP(r *http.Request) {
+   klog.Infof("Received HTTP request, path: %s, method: %s", r.Path, r.Method)
+}
+```
+should be changed to
+```go
+func LogHTTP(r *http.Request) {
+   klog.InfoS("Received HTTP request", "path", r.Path, "method", r.Method)
+}
+```
+
+Names of arguments should use [lowerCamelCase] and be alphanumeric. Arguments names in one log call should be unique.
+Names should be picked based on semantic meaning of value itself, not the context in which is used (log message should
+imply the context). For example names like `status` should be used over (`desiredStatus`, `oldStatus`, `badStatus`) thus
+allowing to query and join different log lines of the `status` field.
+
+Kubernetes objects should be referenced using only their kind, no matter their api group or version. Example argument
+names: `deployment`, `pod`, `node`, `replicaSet`. For objects of unknown type, is ok to log them under `object` key.
+
+In situations when we want to the log value of the same meaning twice (e.g. transition between state) it is ok to name
+an additional argument based on context, but leaving one main value with the name unchanged.
+
+Example:
+
+```go
+func LogStatus(oldStatus, desiredStatus, newStatus string) {
+   klog.InfoS("Changed pod status", "status", newStatus, "oldStatus", oldStatus, "desiredStatus", desiredStatus)
+}
+```
+
+[lowerCamelCase]: https://en.wiktionary.org/wiki/lowerCamelCase
+
+### Use `klog.KObj` and `klog.KRef` for Kubernetes objects
+
+As part of structured logging migration we want to ensure that kubernetes objects references are consistent within the
+codebase. Two new utility functions were introduced to klog `klog.KObj` and `klog.KRef`. Any reference
+(name, uid, namespace) to Kubernetes Object (Pod, Node, Deployment, CRD) should be rewritten to utilize those functions.
+
+For example
+```go
+func updatePod(pod *covev1.Pod) {
+   ...
+   klog.Infof("Updated pod %s in namespace %s", pod.Name, pod.Namespace)
+}
+```
+should be changed to
+```go
+func updatePod(pod *covev1.Pod) {
+   ...
+   klog.InfoS("Updated pod", "pod", klog.KObj(pod))
+}
+```
+And
+```go
+func updatePod(pod *covev1.Pod) {
+   ...
+   klog.Infof("Updated pod with uid: %s", pod.Uid)
+}
+```
+should be changed to
+```go
+func updatePod(pod *covev1.Pod) {
+   ...
+   klog.InfoS("Updated pod", "pod", klog.KObj(pod))
```

Comment:
This loses the UID - is that intentional?  I assume it was added for a reason, and losing it will be a regression

Reply:
One of the nice things we can do with something like this is dynamically filter fields.  I'd suggest we capture UID, and then skip it at lower levels if someone doesn't want it.

Reply:
Proper handling UID was pushed outside scope of initial migration. Currently `klog.KObj` doesn't support UID. Dynamic filtering discussed by @DirectXMan12 is definitely on list of next steps. 

For now we have options:
* Use both `pod` and `pod_uid`. No regression, but can lead to duplication when `klog.KObj` starts using `UID`
* Replace usage of UID with `klog.KObj` by treating `<namespace>/<name>` as equivalent for `UID`. It would work in most places, but in some cases (when using `GenerateName`, or when recreating object where name stays same but UID changes) it will be regression. 

I think option 1 makes sense as it should be possible to scan code base and remove  for usage of `UID` before adding it to `klog.KObj`

Reply:
I am OK with that

Reply:
though Shouldn't it be `podUID` not `pod_uid` ?

Reply:
Yes